doc/monads/transformers.lean

```diff
@@ -0,0 +1,316 @@
+`optionalSwitch` which has type `ReaderT Arguments (Except String) Bool` and call it from
+`parseArguments` which has a bigger type `StateT Config (ReaderT Arguments (Except String))`.
+This "just worked" because Lean did some magic with monad lifting.
+
+To give you a simpler example of this, suppose you have the following funciton:
+-/
+def divide (x : Float ) (y : Float): ExceptT String Id Float :=
+  if y == 0 then
+    throw "can't divide by zero"
+  else
+    pure (x / y)
+
+#eval divide 6 3 -- Except.ok 2.000000
+#eval divide 1 0 -- Except.error "can't divide by zero"
+/-!
+
+Notice here we used the `ExceptT` transformer, but we composed it with the `Id` identity monad.
+This is then the same as writing `Except String Float` since the identity monad does nothing.
+
+Now suppose you want to count the number of times divide is called and store the result in some
+global state:
+-/
+
+def divideCounter (x : Float) (y : Float) : StateT Nat (ExceptT String Id) Float := do
+  modify fun s => s + 1
+  divide x y
+
+#eval divideCounter 6 3 |>.run 0    -- Except.ok (2.000000, 1)
+#eval divideCounter 1 0 |>.run 0    -- Except.error "can't divide by zero"
+
+/-!
+
+The `modify` function is a helper which makes it easier to use `modifyGet` from the `StateM` monad.
+But something interesting is happening here, `divideCounter` is returning the value of
+`divide`, but the types don't match, yet it works?  This is monad lifting in action.
+
+You can see this more clearly with the following test:
+
+-/
+def liftTest (x : Except String Float) :
+  StateT Nat (Except String) Float := x
+
+#eval liftTest (divide 5 1) |>.run 3 -- Except.ok (5.000000, 3)
+
+/-!
+
+Notice that `liftTest` returned `x` without doing anything to it, yet that matched the return type
+`StateT Nat (Except String) Float`.  Monad lifting is provided by monad transformers.  if you
+`#print liftTest` you will see that Lean is implementing this using a call to a function named
+`monadLift` from the `MonadLift` type class:
+
+```lean,ignore
+class MonadLift (m : Type u → Type v) (n : Type u → Type w) where
+  monadLift : {α : Type u} → m α → n α
+```
+
+So `monadLift` is a function for lifting a computation from an inner `Monad m α ` to an outer `Monad n α`.
+You could replace `x` in `liftTest` with `monadLift x` if you want to be explicit about it.
+
+The StateT monad transformer defines an instance of `MonadLift` like this:
+
+```lean
+@[inline] protected def lift {α : Type u} (t : m α) : StateT σ m α :=
+  fun s => do let a ← t; pure (a, s)
+
+instance : MonadLift m (StateT σ m) := ⟨StateT.lift⟩
+```
+This means that any monad `m` can be wrapped in a `StateT` monad by inventing the trivial function
+`fun s => do let a ← t; pure (a, s)` that takes state `s`, runs the inner monad action `t`, and
+returns the result and the new state in a pair `(a, s)` without making any changes to `s`.
+
+Because `MonadLift` is a type class, type inference can automatically find the required `monadLift`
```

Comment:
"type inference can automatically find" 
=>
"Lean can automatically find"

Reply:
Fixed in https://github.com/leanprover/lean4/pull/1505/commits/8dd24bc9dae40eb19b4625bf0b94cfb3b15a2f07